Delete image

wilsonyqm opened the pull request:
Closes #8

bboe commented on test/controllers/images_controller_test.rb:
```diff
@@ -59,4 +59,13 @@ class ImagesControllerTest < ActionController::TestCase
     assert_response :success
     assert_select '#image_tags', text: ''
   end
+
```
What happens if the image you try to delete does not exist?

rmacklin replied:
@wilsonyqm I see that you've added a controller test to test this case. However, I'm curious if you've manually tested this on the review app. Currently the experience when I tried to delete a nonexistent image is that I'm presented with:
<img width="601" alt="screen shot 2016-04-18 at 3 57 59 pm" src="https://cloud.githubusercontent.com/assets/1863540/14622401/8d3784e6-057e-11e6-9e0b-29daa1fd7b48.png">
which is not very user-friendly.

wilsonyqm replied:
I fixed that

bboe commented on app/views/images/index.html.erb:
```diff
@@ -13,6 +13,7 @@
         <td><%= link_to image.title, image_path(image) %></td>
         <td class="test-tags">Tags: <%= image.tag_list %></td>
         <td><%= image_tag image.url, class: 'image' %></td>
+        <td><%= link_to "delete", image, method: :delete, data: { confirm: 'Are you sure?'} %></td>
```
:space_invader: after `'Are you sure?'`

bboe commented on config/routes.rb:
```diff
@@ -1,11 +1,13 @@
 Rails.application.routes.draw do
+  get 'images/destroy'
```
This is redundant to :destroy on line 10.

bboe commented on test/controllers/images_controller_test.rb:
```diff
@@ -59,4 +59,13 @@ class ImagesControllerTest < ActionController::TestCase
+    image = Image.create!(title: 'test4Img', url: image_url, tag_list: '')
+    assert_difference 'Image.count', -1 do
+      delete :destroy, id: image
+    end
```
What is the response status?

rmacklin commented on app/controllers/images_controller.rb:
```diff
@@ -20,6 +20,12 @@ def create
   end
 
+  def destroy
+    Image.find(params[:id]).destroy
```
See https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/27#discussion_r60148954

rmacklin commented on app/views/layouts/application.html.erb:
```diff
@@ -9,9 +9,9 @@
-
     <div class="page-wrapper__content">
       <div class="container m-t-4">
+        <p class="page-wrapper__warning"><%= flash[:success] if flash[:success].present? %></p>
```
This seems a bit strange. We're displaying a "success" message with a red warning/error style.

rmacklin commented on test/controllers/images_controller_test.rb:
```diff
@@ -57,9 +57,25 @@ class ImagesControllerTest < ActionController::TestCase
+
+  test 'delete image does not exist' do
+    assert_no_difference 'Image.count' do
+      assert_raises(ActiveRecord::RecordNotFound) { delete(:destroy, id: -1) }
```
See https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/27#discussion_r60148954

rmacklin commented on test/flow_tests/images_crud_test.rb:
```diff
@@ -26,6 +26,15 @@ class ImagesCrudTest < FlowTestCase
+    image = Image.create!(title: 'delete_test_1', url: 'http://www.horniman.info/DKNSARC/SD04/IMAGES/D4P1570C.JPG')
+    visit(root_path)
+    page.assert_selector("img[src=\"#{image.url}\"]", count: 1)
+    page.find("div[data-image-id=\"#{image.id}\"] a[data-method=\"delete\"]").click
```
Shouldn't this line open a confirmation dialog? How come this test isn't handling that?

rmacklin commented:
Could you please make the style corrections in one commit and combine the rest of the "delete image" changes into another commit?

bboe commented on app/controllers/images_controller.rb:
```diff
@@ -20,7 +22,25 @@ def create
+  def destroy
+    image = Image.find_by(params[:id])
+    if image.present?
+      flash[:success] = 'Image Deleted'
```
Should `deleted` be capitalized?

wilsonyqm replied:
fixed that

bboe commented on app/controllers/images_controller.rb:
```diff
@@ -20,7 +22,25 @@ def create
+      flash[:success] = 'Image Deleted'
+      image.destroy!
+    else
+      flash[:error] = 'Image Not Exist'
```
Consider `Image does not exist`.

wilsonyqm replied:
Fixed

bboe commented on app/controllers/images_controller.rb:
```diff
@@ -20,7 +22,25 @@ def create
+  def render_404
+    respond_to do |format|
+      format.html { render :file => "#{Rails.root}/public/404", :layout => false, :status => :not_found }
+      format.xml { head :not_found }
```
Are you using this format?

wilsonyqm replied:
https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/27#discussion_r60448334

bboe commented on app/controllers/images_controller.rb:
```diff
@@ -20,7 +22,25 @@ def create
+    respond_to do |format|
+      format.html { render :file => "#{Rails.root}/public/404", :layout => false, :status => :not_found }
+      format.xml { head :not_found }
+      format.any { head :not_found }
```
Are you using any format other than html?

wilsonyqm replied:
Nope, it's just been used for debugging the exceptions, forgot to delete it.

bboe commented on app/views/layouts/application.html.erb:
```diff
@@ -9,9 +9,14 @@
-
     <div class="page-wrapper__content">
       <div class="container m-t-4">
+        <p class="page-wrapper__success">
```
Can you add the condition to this tag so that an empty `<p>` tag isn't always on the page?

wilsonyqm replied:
fixed on the master branch with @rmacklin  and merged [here](https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/blob/deleteImage/app/views/layouts/application.html.erb#L15)

bboe commented on app/views/layouts/application.html.erb:
```diff
@@ -9,9 +9,14 @@
+        <p class="page-wrapper__success">
+          <%= flash[:success] if flash[:success].present? %>
+        </p>
+        <p class="page-wrapper__warning">
```
Can you add the condition to this tag so that an empty `<p>` tag isn't always on the page?

wilsonyqm replied:
see https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/27#discussion_r60677977

bboe commented on test/controllers/images_controller_test.rb:
```diff
@@ -57,9 +57,26 @@ class ImagesControllerTest < ActionController::TestCase
+  test 'delete image does not exist' do
+    assert_no_difference 'Image.count' do
+      delete(:destroy, id: -1)
+      assert_equal 'Image Not Exist' ,flash[:error]
```
Does this assertion need to be inside of the `assert_no_difference` block? Also have you run pronto on these changes?

wilsonyqm replied:
I am not so sure what will happen if there is a nested assertion, but I think put it outside would be definitely better

wilsonyqm replied:
The response status is 200, basically it just pops out a warning message and redirects to images index page

bboe replied:
I asked about the response status because it is not being tested in this test.

bboe commented:
Making great progress. I added a few comments that would be great to address.

bboe commented on app/views/layouts/application.html.erb:
```diff
@@ -8,14 +8,17 @@
       <div class="container m-t-4">
+        <% if flash[:success].present? %>
+          <p class="page-wrapper__success"><%= flash[:success] %></p>
+        <% elsif flash[:error].present? %>
```
Do you think it ever possible that there could be both a success and error flash message at the same time? Maybe there is not, but just something to think about based on how this code is written.

bboe commented on test/controllers/images_controller_test.rb:
```diff
@@ -57,9 +57,26 @@ class ImagesControllerTest < ActionController::TestCase
+  test 'delete image does not exist' do
+    assert_no_difference 'Image.count' do
+      delete(:destroy, id: -1)
+      assert_equal 'Image does not exist', flash[:error]
```
Does this assertion need to be within the `assert_no_difference` block?

wilsonyqm replied:
Fixed here https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/27/commits/5c684f640630bdd2d6c046c9d8fc611fdca5159e#diff-8807b315bee3099305dbdee3d4f0a4f7R80

bboe commented on test/flow_tests/images_crud_test.rb:
```diff
@@ -26,6 +26,18 @@ class ImagesCrudTest < FlowTestCase
+    page.assert_selector("img[src=\"#{image2.url}\"]", count: 0)
+    page.assert_selector("img[src=\"#{image1.url}\"]", count: 1)
+    page.assert_text('Image deleted')
   end
```
Can you add a flow test for attempting to delete an image that does not exist? There's a small trick to accomplish that, which I'll let you ponder first. If you're not sure, chat with Xin as she recently did similar.

wilsonyqm replied:
Fixed with image.destroy!   https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/27/commits/5c684f640630bdd2d6c046c9d8fc611fdca5159e

bboe commented:
A few comments, and it appears that a rebase is required for this to be merged. Once those are addressed this story should be good to go. Awesome work.

wilsonyqm commented:
I Fixed all the code mentioned in the comments, please take a look at it.

rmacklin commented on app/assets/stylesheets/_page-wrapper.scss:
```diff
@@ -7,3 +7,13 @@
+.page-wrapper__warning {
+  color: red;
+  font-size: 30px;
+}
```
These classes are not used - please remove them. (I believe they used to be used, until you rebased onto the latest changes, which made these classes obsolete)

wilsonyqm replied:
Removed

rmacklin commented on app/controllers/images_controller.rb:
```diff
@@ -1,4 +1,6 @@
 class ImagesController < ApplicationController
+  include ActiveSupport::Rescuable
```
Why was this added?

wilsonyqm replied:
This was added to catch the exception when image does not exist, but we fixed it in other way, so It should be deleted.

rmacklin replied:
It shouldn't have been needed even using the other way.

rmacklin commented on test/controllers/images_controller_test.rb:
```diff
@@ -57,9 +57,26 @@ class ImagesControllerTest < ActionController::TestCase
+    assert_no_difference 'Image.count' do
+      delete(:destroy, id: -1)
+    end
+    assert_equal 'Image does not exist', flash[:error]
```
What can we assert about the response?

wilsonyqm replied:
When I did the assert_response, the status shows it is :402, which is found HTTP status. It should be :not_found status, right? I will assert_redirected in this case

wilsonyqm replied:
https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/27/commits/143936fb8265fe561fd6f10b247a91259c43ba7d#diff-8807b315bee3099305dbdee3d4f0a4f7R80

rmacklin replied:
402? I think you mean 302. That is because we're redirecting in this case. We're _not_ returning `:not_found`.

wilsonyqm replied:
Right, 302 Found.

rmacklin commented on test/flow_tests/images_crud_test.rb:
```diff
@@ -26,6 +26,33 @@ class ImagesCrudTest < FlowTestCase
+  end
+
+  test 'delete an image' do
+
```
Random newline here

wilsonyqm replied:
Fixed.

rmacklin commented on test/flow_tests/images_crud_test.rb:
```diff
@@ -26,6 +26,33 @@ class ImagesCrudTest < FlowTestCase
+  test 'delete an image' do
+
+    image1 = Image.create!(title: 'delete_test_1', url: 'http://www.horniman.info/DKNSARC/SD04/IMAGES/D4P1570C.JPG')
+    image2 = Image.create!(title: 'delete_test_2', url: 'https://upload.wikimedia.org/wikipedia/commons/9/92/Colorful_spring_garden.jpg')
```
Consider calling this variable something like `image_to_delete`

wilsonyqm replied:
Fixed

rmacklin replied:
I'm not sure why you renamed both variables. It doesn't make sense in the test. See https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/27#discussion_r60771322

wilsonyqm replied:
Fixed and renamed the one that we need to delete

rmacklin commented on test/flow_tests/images_crud_test.rb:
```diff
@@ -26,6 +26,33 @@ class ImagesCrudTest < FlowTestCase
+
+  test 'delete an not exist image' do
+    image1 = Image.create!(title: 'delete_test_1', url: 'http://www.horniman.info/DKNSARC/SD04/IMAGES/D4P1570C.JPG')
+    image2 = Image.create!(title: 'delete_test_2', url: 'https://upload.wikimedia.org/wikipedia/commons/9/92/Colorful_spring_garden.jpg')
```
Do we need to create two images for this test?

wilsonyqm replied:
Yes, I think we need one extra image as the comparison to make sure this destroy is working on the right image one.

rmacklin replied:
We are not actually destroying any images in this test. That argument only makes sense for the previous test where we actually destroy an image.

wilsonyqm replied:
Fixed.

rmacklin commented on test/flow_tests/images_crud_test.rb:
```diff
@@ -26,6 +26,33 @@ class ImagesCrudTest < FlowTestCase
+    visit(root_path)
+    page.assert_selector("img[src=\"#{image2.url}\"]", count: 1)
+    page.assert_selector("img[src=\"#{image1.url}\"]", count: 1)
+    page.find("div[data-image-id=\"#{image2.id}\"] a[data-method=\"delete\"]").click
```
This should bring up a confirmation dialog. We should handle that here. I believe the reason this works is that you're not using a javascript-capable Capybara driver. You can configure the `default_driver` to be `Capybara.javascript_driver` this your flow_test_helper. After doing that, you should be able to test the confirmation dialog.

wilsonyqm replied:
Add <Capybara.default_driver = :selenium> as the environment

wilsonyqm replied:
add accept_confirm. https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/27/commits/143936fb8265fe561fd6f10b247a91259c43ba7d#diff-ada26b2f9a3d6561f73ea5b0933d7ff3R38

rmacklin replied:
See https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/27#discussion_r60770740 and https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/27#discussion_r60771230

wilsonyqm replied:
Fixed, and add dismiss/accept confirm with text assertion

rmacklin commented:
> I Fixed all the code mentioned in the comments, please take a look at it.

Thanks. I added some new comments.

rmacklin commented on test/flow_test_helper.rb:
```diff
@@ -24,3 +24,4 @@ def self.connection
 # Forces all threads to share the same connection. This works on
 # Capybara because it starts the web server in a thread.
 ActiveRecord::Base.shared_connection = ActiveRecord::Base.connection
+Capybara.default_driver = :selenium
```
You should set this to `Capybara.javascript_driver`, rather than explicitly to `:selenium`. Also, please put a newline before this line because it's not part of the code above this which has a comment explaining it.

bboe replied:
@wilsonyqm did you see this comment?

wilsonyqm replied:
Yes, I fixed it in this new commit.
https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/27/commits/61306c86623c2177b080b17a6ba228ef0e15d653

rmacklin commented on test/flow_tests/images_crud_test.rb:
```diff
@@ -30,29 +30,32 @@ class ImagesCrudTest < FlowTestCase
-
-    image1 = Image.create!(title: 'delete_test_1', url: 'http://www.horniman.info/DKNSARC/SD04/IMAGES/D4P1570C.JPG')
-    image2 = Image.create!(title: 'delete_test_2', url: 'https://upload.wikimedia.org/wikipedia/commons/9/92/Colorful_spring_garden.jpg')
+    image_to_delete1 = Image.create!(title: 'delete_test_1', url: 'http://www.horniman.info/DKNSARC/SD04/IMAGES/D4P1570C.JPG')
```
Why did you also rename this? Only one image should be called `image_to_delete` because you are only deleting one image.

wilsonyqm commented:
Fixed all the bugs. And the flow test is awesome now, which will pop up a firefox browser, it's cool!

rmacklin commented on test/flow_tests/images_crud_test.rb:
```diff
@@ -30,29 +30,32 @@ class ImagesCrudTest < FlowTestCase
-    page.assert_selector("img[src=\"#{image1.url}\"]", count: 1)
+    page.assert_selector("img[src=\"#{image_to_delete2.url}\"]", count: 1)
+    page.assert_selector("img[src=\"#{image_to_delete1.url}\"]", count: 1)
+    accept_confirm do
```
You should also test the `dismiss_confirm` case. Furthermore, you should check the content of the confirmation dialog.

rmacklin commented on test/flow_tests/images_crud_test.rb:
```diff
@@ -26,6 +26,36 @@ class ImagesCrudTest < FlowTestCase
+    visit(root_path)
+    page.assert_selector("img[src=\"#{image_to_delete2.url}\"]", count: 1)
+    page.assert_selector("img[src=\"#{image_to_delete1.url}\"]", count: 1)
+    accept_confirm do
```
You should also test the `dismiss_confirm` case. Furthermore, you should check the content of the confirmation dialog.

wilsonyqm replied:
Is this what we usually do for dealing with dialog confirmation? Fixed now

rmacklin commented on test/flow_tests/images_crud_test.rb:
```diff
@@ -26,6 +26,36 @@ class ImagesCrudTest < FlowTestCase
+  end
+
+  test 'delete an image' do
+    image_to_delete1 = Image.create!(title: 'delete_test_1', url: 'http://www.horniman.info/DKNSARC/SD04/IMAGES/D4P1570C.JPG')
```
Why did you also rename this? Only one image should be called `image_to_delete` because you are only deleting one image.

wilsonyqm replied:
Sorry, I misunderstood you. Now it's been fixed

rmacklin commented on app/controllers/images_controller.rb:
```diff
@@ -1,4 +1,5 @@
 class ImagesController < ApplicationController
+
```
Please remove this extra newline

wilsonyqm replied:
Fixed

rmacklin commented on test/flow_tests/images_crud_test.rb:
```diff
@@ -26,6 +26,36 @@ class ImagesCrudTest < FlowTestCase
+
+  test 'delete an not exist image' do
+    image_to_delete1 = Image.create!(title: 'delete_test_1', url: 'http://www.horniman.info/DKNSARC/SD04/IMAGES/D4P1570C.JPG')
+    image_to_delete2 = Image.create!(title: 'delete_test_2', url: 'https://upload.wikimedia.org/wikipedia/commons/9/92/Colorful_spring_garden.jpg')
```
See https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/27#discussion_r60770948. We should not need to create two images for this test.

wilsonyqm replied:
Deleted other image

rmacklin commented on test/controllers/images_controller_test.rb:
```diff
@@ -57,9 +57,27 @@ class ImagesControllerTest < ActionController::TestCase
+
+  test 'delete image does not exist' do
+    assert_no_difference 'Image.count' do
+      delete(:destroy, id: -1)
```
Why did you use parentheses here?

wilsonyqm replied:
Fixed

rmacklin commented on test/flow_tests/images_crud_test.rb:
```diff
@@ -26,6 +26,36 @@ class ImagesCrudTest < FlowTestCase
+    end
+    page.assert_selector("img[src=\"#{image_to_delete2.url}\"]", count: 0)
+    page.assert_selector("img[src=\"#{image_to_delete1.url}\"]", count: 1)
+    page.assert_text('Image deleted')
```
Can we be more specific about where we expect this text to appear?

wilsonyqm replied:
add 
```ruby
page.assert_current_path(images_path)
```

rmacklin replied:
That's not what I meant. See https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/27#discussion_r60946336

rmacklin commented on test/flow_tests/images_crud_test.rb:
```diff
@@ -26,6 +26,36 @@ class ImagesCrudTest < FlowTestCase
+    accept_confirm do
+      page.find("div[data-image-id=\"#{image_to_delete2.id}\"] a[data-method=\"delete\"]").click
+    end
+    page.assert_selector(".page-wrapper__content", {text: 'Image does not exist', count: 1})
```
Can we be more specific than `.page-wrapper__content` about where we expect this text to appear?

wilsonyqm replied:
See https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/27#discussion_r60797153

rmacklin replied:
Again, not what I meant. I'm talking about using a more specific selector, because almost everything on the page is inside `.page-wrapper__content`

rmacklin commented:
A new round of comments has been added.

wilsonyqm commented:
New comments have been addressed. https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/27/commits/53cec73f28ae22eb502c52a0ca4731c28750fc79

rmacklin commented on test/flow_tests/images_crud_test.rb:
```diff
@@ -26,6 +26,49 @@ class ImagesCrudTest < FlowTestCase
+    page.assert_current_path(images_path)
+    page.assert_selector("img[src=\"#{image_to_delete.url}\"]", count: 0)
+    page.assert_selector("img[src=\"#{image1.url}\"]", count: 1)
+    page.assert_text('Image deleted')
```
Can we be more specific about where this text should appear?

wilsonyqm replied:
Fixed with `page.assert_selector('.alert-success', {count: 1, text: 'Image deleted' } )`

rmacklin replied:
That's good. Another thing we could do is wrap [this](https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/blob/61306c86623c2177b080b17a6ba228ef0e15d653/app/views/layouts/application.html.erb#L15-L21) in a `div` with the id `flash_messages`, and then use `#flash_messages .alert-success` for the selector.

Additionally, it's idiomatic to remove the curly braces around a hash if it is the last argument of a method.

wilsonyqm replied:
It seems like I cannot use both of the #flash_messages .alert-success as the selector, since they are in the same div. So I will use #flashe_messages.

rmacklin replied:
I explicitly said not to make them the same div:

> wrap [this](https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/blob/61306c86623c2177b080b17a6ba228ef0e15d653/app/views/layouts/application.html.erb#L15-L21) in a `div` with the id `flash_messages`

The change you made [here](https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/commit/e51c6e5785ce1092474bb66c1095ad59c6e53132#diff-9599427925097c3c66f26ac1e0de5cadL17) is incorrect, because it can produce multiple elements with the same `id`, which is invalid HTML.

rmacklin commented on test/flow_tests/images_crud_test.rb:
```diff
@@ -26,6 +26,49 @@ class ImagesCrudTest < FlowTestCase
+    image_to_delete.destroy!
+    page.assert_selector("img[src=\"#{image_to_delete.url}\"]", count: 1)
+
+    dismiss_confirm('Are you sure') do
```
I don't think it's necessary to test dismissing the dialog a second time.

wilsonyqm replied:
Deleted

rmacklin commented on test/flow_tests/images_crud_test.rb:
```diff
@@ -26,6 +26,49 @@ class ImagesCrudTest < FlowTestCase
+      page.find("div[data-image-id=\"#{image_to_delete.id}\"] a[data-method=\"delete\"]").click
+    end
+    page.assert_current_path(images_path)
+    page.assert_selector(".page-wrapper__content", {text: 'Image does not exist', count: 1})
```
Is there a more specific selector we can use for where we expect this message to appear?

wilsonyqm replied:
Fixed with `page.assert_selector('.page-wrapper__content .alert-danger', {text: 'Image does not exist', count: 1})`

rmacklin replied:
As mentioned in https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/27#discussion_r60951013, I think a better selector would be `#flash_messages .alert-danger`. Also, we should remove the curly braces here.

rmacklin commented:
Do you think users should also be able to delete images from the show page?

rmacklin commented:
This is looking a lot better. Just a few comments remain.

wilsonyqm commented:
add delete button on show page, and fixed the style when image does not exist. Please take a look when you have time. Thanks

rmacklin commented on app/views/images/show.html.erb:
```diff
@@ -6,5 +6,6 @@
 </div>
-
+<%= link_to 'Delete', image_path(@image), method: :delete, 'data-confirm': 'Are you sure?',
+            class: 'btn btn-danger'%>
```
As asked in https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/28#issuecomment-214440571, is there a way we can refactor the code such that it's impossible for the index and show page to get out of sync for things like this?

wilsonyqm replied:
In this case, if you click on delete image on the show page, the controller will decide whether this image exist and return the alert. In the other hand, if the image has been deleted, and the index has not been refreshed (image still exist), user clicks on the image show page, and this would lead to an error, probably we need a if condition on the show page controller like this: 
```ruby
def show
    @image = Image.find_by(id: params[:id])
    if @image.nil?
      flash[:danger] = 'Image does not exist'
      redirect_to images_path
    end
  end
```

rmacklin replied:
I don't think you understood my question.

You previously added a delete button on the index page. Then you added on one the show page. I'm asking: is there a way you can refactor the code such that you only need to make _one change_ and then the delete button will appear on both the index page and the show page?

wilsonyqm replied:
I think I understand your question, but I cannot find a way to implement that, any hint on this?

rmacklin replied:
Can you explain your thought process so far?

wilsonyqm replied:
So I would like to use a render partial, so that image component is combined with delete button and image itself including the image title and tags. So I will initial a image.html.erb file, and it will be rendered in both index and show page.

rmacklin replied:
Cool. This is an example of https://en.wikipedia.org/wiki/Don%27t_repeat_yourself, which you should always keep in mind.

rmacklin commented on test/flow_tests/images_crud_test.rb:
```diff
@@ -48,7 +48,7 @@ class ImagesCrudTest < FlowTestCase
     page.assert_selector("img[src=\"#{image_to_delete.url}\"]", count: 0)
     page.assert_selector("img[src=\"#{image1.url}\"]", count: 1)
-    page.assert_text('Image deleted')
+    page.assert_selector('.page-wrapper__content .alert-success', {count: 1, text: 'Image deleted' } )
```
See https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/27#discussion_r60951013

rmacklin commented on test/flow_tests/images_crud_test.rb:
```diff
@@ -57,18 +57,11 @@ class ImagesCrudTest < FlowTestCase
     end
     page.assert_current_path(images_path)
-    page.assert_selector(".page-wrapper__content", {text: 'Image does not exist', count: 1})
+    page.assert_selector('.page-wrapper__content .alert-danger', {text: 'Image does not exist', count: 1})
```
See https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/27#discussion_r60951143

rmacklin commented on test/flow_tests/images_crud_test.rb:
```diff
@@ -26,6 +26,42 @@ class ImagesCrudTest < FlowTestCase
+      page.find("div[data-image-id=\"#{image_to_delete.id}\"] a[data-method=\"delete\"]").click
+    end
+    page.assert_current_path(images_path)
+    page.assert_selector('.page-wrapper__content .alert-danger', {text: 'Image does not exist', count: 1})
```
See https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/27#discussion_r60951143

rmacklin commented on test/flow_tests/images_crud_test.rb:
```diff
@@ -26,6 +26,42 @@ class ImagesCrudTest < FlowTestCase
+    page.assert_current_path(images_path)
+    page.assert_selector("img[src=\"#{image_to_delete.url}\"]", count: 0)
+    page.assert_selector("img[src=\"#{image1.url}\"]", count: 1)
+    page.assert_selector('.page-wrapper__content .alert-success', {count: 1, text: 'Image deleted' } )
```
See https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/27#discussion_r60951013

wilsonyqm commented:
I have addressed your comment https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/27#discussion_r60955239  and discussed with Bryce. I initialize a image.html.erb and call render partial in show and index page. So it would be easy to modify later.

rmacklin commented on app/views/layouts/application.html.erb:
```diff
@@ -14,7 +14,7 @@
         <% flash.each do |type, message| %>
           <% if flash[type].present? %>
-            <div class="alert alert-<%= type %>">
+            <div id="flash_messages" class="alert alert-<%= type %>">
```
See https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/27#discussion_r60975583

wilsonyqm replied:
Fixed

rmacklin commented on app/views/images/show.html.erb:
```diff
@@ -1,10 +1,2 @@
-  <%= image_tag @image.url, class: 'image-detail__image' %>
-</div>
-
+<%= render partial: 'images/image', locals: {image: @image} %>
```
Does `render @image` work here?

wilsonyqm replied:
Yep, it works, and this way of implementation is mentioned in the rails tutorial. http://guides.rubyonrails.org/layouts_and_rendering.html#using-partials

rmacklin commented on app/views/images/index.html.erb:
```diff
@@ -5,10 +5,8 @@
-        <h2 class="card-title image__title"><%= link_to image.title, image_path(image) %></h2>
-        <h2 class="test-tags image__title">Tags: <%= image.tag_list %></h2>
+      <div class="card-block image img-fluid", data-image-id=<%= image.id %>>
+        <%= render partial: 'images/image', locals: {image: image} %>
```
Does `render image` work here?

wilsonyqm replied:
Yes. It works

rmacklin commented on app/controllers/images_controller.rb:
```diff
@@ -4,7 +4,11 @@ def index
+    if @image.nil?
+      flash[:danger] = 'Image does not exist'
+      redirect_to images_path
+    end
```
Please revert this. This change was only made as a misunderstanding of this comment: https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/27#discussion_r60951342

wilsonyqm replied:
reverted

rmacklin commented on app/views/images/_image.html.erb:
```diff
@@ -0,0 +1,9 @@
+  <p class="image-detail__label">Tags:
+    <span id="image_tags" class="image-detail__tags"><%= image.tag_list %></span>
+  </p>
+  <%= image_tag image.url, class: 'image__box' %>
```
You've broken BEM here, because this partial is rendered once where there is no ancestor with the `image` class (making the `image__box` class invalid). Please work with @aaronpanch to resolve this.

wilsonyqm replied:
Revised in code.

rmacklin commented on app/views/images/_image.html.erb:
```diff
@@ -0,0 +1,9 @@
+<div class="image-detail">
+  <p id="image_title"><%= link_to image.title, image_path(image), class: 'image__title' %></p>
```
No elements in this partial should have an `id` because it will create duplicate `id`s when this partial is rendered multiple times on the index page.

wilsonyqm replied:
That makes sense, and it really matters in the partial file. Great point.

rmacklin commented on app/views/images/_image.html.erb:
```diff
@@ -0,0 +1,9 @@
+<div class="image-detail">
+  <p id="image_title"><%= link_to image.title, image_path(image), class: 'image__title' %></p>
+  <p class="image-detail__label">Tags:
+    <span id="image_tags" class="image-detail__tags"><%= image.tag_list %></span>
```
Again, do not use an `id` in this partial

wilsonyqm replied:
Got you.

rmacklin commented:
> I initialize a image.html.erb and call render partial in show and index page.

Great. This is an example of https://en.wikipedia.org/wiki/Don%27t_repeat_yourself, which you should always keep in mind.

wilsonyqm commented:
Addressed comments and well organized the css BEM, https://github.com/AppFolioOnboarding/image-sharer-wilsonyqm/pull/27/commits/7c30892a9ad013c6736e6ff9ad2762fbedb48fdc

rmacklin commented:
Looks good

rmacklin commented:
Woops. I forgot that you hadn't squashed the commits. Oh well. Please rebase the other branch now.
